applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -919,6 +929,116 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::ComputePotentia
     }
 }
 
+template <int TDim, int TNumNodes>
+void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::FindUpwindElement(const ProcessInfo& rCurrentProcessInfo)
+{
+    const TransonicPerturbationPotentialFlowElement& r_this = *this;
+
+    // current element geometry
+    const GeometryType& r_geom = r_this.GetGeometry();
+    const Point current_element_center = r_geom.Center();
+
+    // current element edges
+    const auto edges = r_geom.GenerateEdges();
+
+    // get local coordinates of element center
+    array_1d<double,3> aux_coordinates;
+    r_geom.PointLocalCoordinates(aux_coordinates, current_element_center);
+
+    // outward pointing normals of each edge
+    auto first_edge_normal = edges[0].Normal(aux_coordinates);
+    auto second_edge_normal = edges[1].Normal(aux_coordinates);
+    auto third_edge_normal = edges[2].Normal(aux_coordinates);
+
+    // make normals point into element
+    first_edge_normal[0] = -1.0 * first_edge_normal[0];
+    first_edge_normal[1] = -1.0 * first_edge_normal[1];
+
+    second_edge_normal[0] = -1.0 * second_edge_normal[0];
+    second_edge_normal[1] = -1.0 * second_edge_normal[1];
+
+    third_edge_normal[0] = -1.0 * third_edge_normal[0];
+    third_edge_normal[1] = -1.0 * third_edge_normal[1];
+
+    // get free stream velocity and norm
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    const double free_stream_velocity_norm = MathUtils<double>::Norm3(free_stream_velocity);
+
+    // find component of element edge normal vector in direction of free stream velocity
+    const double first_edge_normal_fs_comp = inner_prod(first_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+    const double second_edge_normal_fs_comp = inner_prod(second_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+    const double third_edge_normal_fs_comp = inner_prod(third_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+
+    vector<double> edge_normal_velocity_components (3);
+    edge_normal_velocity_components[0] = first_edge_normal_fs_comp;
+    edge_normal_velocity_components[1] = second_edge_normal_fs_comp;
+    edge_normal_velocity_components[2] = third_edge_normal_fs_comp;
+
+    this->SetValue(EDGE_NORMAL_VELOCITY_COMPONENTS, edge_normal_velocity_components);
```

Comment:
We need to think of a way of testing this without storing an extra variable. Maybe bring the functions that do not access private members to the utilities so that you can access them from your tests?